extension/ErrorReporting.jsm

```diff
@@ -6,83 +6,85 @@ const {utils: Cu, interfaces: Ci, classes: Cc} = Components;
-  "chrome javascript",
-  "chrome registration",
-  "XBL",
-  "XBL Prototype Handler",
-  "XBL Content Sink",
-  "xbl javascript",
-  "FrameConstructor",
-];
-
-const listeners = new Set();
+let fingerprintsToReport = [];
+let interval = null;
 
 this.ErrorReporting = {
   startup() {
-    Services.console.registerListener(ErrorReporting);
+    // Submit any collected errors once an hour
+    interval = setInterval(
+      this.submitErrors.bind(this),
+      Services.prefs.getIntPref(
+        "extensions.shield-study-js-errors@shield.mozilla.org.submitIntervalMs",
+        1000 * 60 * 60, // 1 hour
+      ),
+    );
   },
 
   shutdown() {
-    Services.console.unregisterListener(ErrorReporting);
-  },
+    if (interval !== null) {
+      clearInterval(interval);
+    }
 
-  addListener(listener) {
-    listeners.add(listener);
+    // Submit any unsent intervals when the add-on or browser is shutting down
+    this.submitErrors();
   },
 
-  removeListener(listener) {
-    listeners.delete(listener);
+  /**
+   * Create an anonymized fingerprint of the error using the name, message, and
+   * stack trace to create a SHA384 hash.
+   * @param {Error} error
+   * @return {String}
+   */
+  generateFingerprint(error) {
+    const dataToHash = new TextEncoder().encode(
+       [error.name, error.message, error.stack].join(":"),
+     );
+     const hasher = Cc["@mozilla.org/security/hash;1"].createInstance(Ci.nsICryptoHash);
+     hasher.init(hasher.SHA384);
+     hasher.update(dataToHash, dataToHash.length);
+     return hasher.finish(true).replace(/\+/g, "-").replace(/\//g, "_");
   },
 
-  observe(message) {
-    try {
-      message.QueryInterface(Ci.nsIScriptError);
-    } catch (err) {
-      // Not an error, which is fine.
-      return;
-    }
-
-    if (!CATEGORY_WHITELIST.includes(message.category)) {
-      return;
-    }
-
-    let errorMessage = message.errorMessage;
-    let errorName = "Error";
-    if (message.errorMessage.match(ERROR_PREFIX_RE)) {
-      const parts = message.errorMessage.split(":");
-      errorName = parts[0];
-      errorMessage = parts.slice(1).join(":");
-    }
-
-    const error = new Error(errorMessage, message.sourceName, message.lineNumber);
-    error.name = errorName;
-    if (message.stack) {
-      let stackString = "";
-      let s = message.stack;
-      while (s !== null) {
-        stackString += `${s.functionDisplayName}@${s.source}:${s.line}:${s.column}\n`;
-        s = s.parent;
-      }
-      error.stack = stackString;
-    }
-
-    for (const listener of listeners) {
-      try {
-        listener(error);
-      } catch (err) {}
+  submitErrors() {
+    if (fingerprintsToReport.length > 0) {
+      const payload = {
+        version: 3,
+        study_name: "shield-study-js-errors",
+        branch: "default",
+        addon_version: "0.1.0",
+        shield_version: "unset",
+        type: "shield-study-addon",
+        data: {
+          fingerprints: fingerprintsToReport,
```

Comment:
I think that the shield-study-addon schema requires all keys to be strings. You may need to `JSON.stringify()` this, which has worked fine on previous studies.